Reorder arguments to stream readers and writers.

cczhu opened the pull request:
Also reordered reader/writer docstrings, repr functions, and arguments of some
payload methods.

Addresses #112.

mhvk commented on baseband/gsb/base.py:
```diff
@@ -522,49 +526,55 @@ def open(name, mode='rs', **kwargs):
+        give keyword arguments to construct a header.
     **kwargs
         If the header is not given, an attempt will be made to construct one
         with any further keyword arguments.  If one requires to explicitly set
```
Maybe follow other formats and have a line with
```
--- Header Keywords ...
```
I would only put `header_mode` and `time` there...

cczhu replied:
The only format that has `Header Keyworlds` in `open` is DADA, because the header is built in the opener rather than in `DADAStreamWriter`.

mhvk commented on baseband/gsb/frame.py:
```diff
@@ -36,15 +36,15 @@ class GSBFrame(VLBIFrameBase):
     def fromfile(cls, fh_ts, fh_raw, payloadsize=1 << 24, nchan=1, bps=4,
                  complex_data=False, valid=True, verify=True):
-        """Read a frame from timestamp and raw data file handles.
+        """Read a frame from timestamp and raw data filehandles.
```
I think removing the space is OK, though note that the formal python term for this appears to be "file object"...

cczhu replied:
This was inconsistent throughout the docstrings (which used to either use `filehandle` or `file handle`), so I just removed whichever was less popular.

mhvk commented on baseband/gsb/tests/test_gsb.py:
```diff
@@ -229,7 +229,7 @@ def test_payload(self):
             assert payload1.sample_shape == (1,)
             assert payload1.sample_shape.nchan == 1
-            # Passing ``None`` to `payloadsize`throws an error.
+            # Passing `None` to ``payloadsize`` throws an error.
```
This is actually a rather useless comment! If we keep anything: `Reader needs to be passed an actual payloadsize.`

mhvk commented on baseband/mark4/base.py:
```diff
@@ -36,7 +36,7 @@ def read_frame(self, ntrack, decade=None, ref_time=None):
         ref_time : `~astropy.time.Time`, or None, optional
             Reference time within 4 years of the observation time.  Used only
-            if `decade` is ``None``.
+            if `decade` is `None`.
```
maybe just `is not given`?

cczhu replied:
I'm not certain about this.  If the user explicitly sets `decade=None` the same behaviour would result.

mhvk commented on baseband/mark4/base.py:
```diff
@@ -189,22 +189,21 @@ def find_header(self, template_header=None, ntrack=None, decade=None,
-            Number of tracks used to store the data.  Required if
-            ``template_header`` is ``None``.
+            Number of Mark 4 bitstreams.  Required if ``template_header`` is
+            `None`.
```
`not given`

mhvk commented on baseband/mark4/base.py:
```diff
@@ -189,22 +189,21 @@ def find_header(self, template_header=None, ntrack=None, decade=None,
-            ambiguity in the Mark 4 time stamp.  Required if
-            ``template_header`` is ``None``.
+            Decade in which the observations were taken.  Required if
+            ``template_header`` is `None`.
```
not given

mhvk commented on baseband/mark4/base.py:
```diff
@@ -245,40 +244,40 @@ def write_frame(self, data, header=None, **kwargs):
-    fh_raw : `~baseband.mark4.Mark4FileReader`
-        File handle to the raw Mark 4 data stream.
+    fh_raw : filehandle
+        filehandle of the raw Mark 4 stream.
```
Capitalize

mhvk commented on baseband/mark4/base.py:
```diff
@@ -245,40 +244,40 @@ def write_frame(self, data, header=None, **kwargs):
     ntrack : int, or None
-        Number of tracks used to store the data.  If ``None``, will attempt to
-        automatically detect it by scanning the file.
+        Number of Mark 4 bitstreams.  If `None`, will attempt to automatically
```
would write `optional` above, and "If not given" here

cczhu replied:
For consistency with the above, shouldn't we remove all `optional` from header keywords?

mhvk commented on baseband/mark4/base.py:
```diff
@@ -245,40 +244,40 @@ def write_frame(self, data, header=None, **kwargs):
     ref_time : `~astropy.time.Time`, or None, optional
         Reference time within 4 years of the start time of the observations.
-        Used only if `decade` is ``None``.
+        Used only if `decade` is `None`.
```
not given

mhvk commented on baseband/mark4/base.py:
```diff
@@ -513,39 +517,40 @@ def write(self, data, invalid_data=False):
 ntrack : int, optional
-    Number of tracks used to store the data. Will be inferred from file
-    if not given.
+    Number of Mark 4 bitstreams.  If `None`, will attempt to automatically
```
not given

mhvk commented on baseband/mark4/base.py:
```diff
@@ -513,39 +517,40 @@ def write(self, data, invalid_data=False):
-    only if `decade` is ``None``.
-subset : slice or other indexing object, optional
-    Specific components (ie. channels) of the complete sample to decode.  By
+    only if `decade` is `None`.
```
not given

mhvk commented on baseband/mark4/base.py:
```diff
@@ -513,39 +517,40 @@ def write(self, data, invalid_data=False):
     decoded data.
 
---- For writing a stream : (see `~baseband.mark4.base.Mark4StreamWriter`)
+--- For writing a stream: (see :class:`~baseband.mark4.base.Mark4StreamWriter`)
```
I think the space was actually needed for much nicer typesetting (as it makes it look like an argument -- please try; obviously, this affects all formats).

cczhu replied:
If the space is left in the line goes to 80 characters, which is why I originally removed it.  I'll remove `:class:` instead (which I never cared for, but does make the docstrings inconsistent).

mhvk commented on baseband/mark4/frame.py:
```diff
@@ -114,7 +114,7 @@ def fromfile(cls, fh, ntrack, decade=None, ref_time=None, verify=True):
         ref_time : `~astropy.time.Time`, or None, optional
             Reference time within 4 years of the observation time.  Used only
-            if `decade` is ``None``.
+            if `decade` is `None`.
```
not given

mhvk commented on baseband/mark4/header.py:
```diff
@@ -253,7 +253,7 @@ class Mark4Header(Mark4TrackHeader):
     ref_time : `~astropy.time.Time`, or None, optional
         Reference time within 4 years of the observation time, used to infer
-        the full Mark 4 timestamp.  Used only if `decade` is ``None``.
+        the full Mark 4 timestamp.  Used only if `decade` is `None`.
```
not given

mhvk commented on baseband/mark4/header.py:
```diff
@@ -377,10 +377,10 @@ def fromfile(cls, fh, ntrack, decade=None, ref_time=None, verify=True):
         ref_time : `~astropy.time.Time`, or None, optional
             Reference time within 4 years of the observation time.  Used only
-            if `decade` is ``None``.
+            if ``decade`` is `None`.
```
not given

mhvk commented on baseband/mark4/header.py:
```diff
@@ -417,10 +417,10 @@ def fromvalues(cls, ntrack, decade=None, ref_time=None, **kwargs):
         ref_time : `~astropy.time.Time`, or None, optional
             Reference time within 4 years of the observation time.  Not needed
-            if ``time`` is given, and used only if `decade` is ``None``.
+            if ``time`` is given, and used only if `decade` is `None`.
```
not given

mhvk commented on baseband/mark5b/base.py:
```diff
@@ -58,21 +58,21 @@ def find_header(self, template_header=None, framesize=None, kday=None,
         framesize : int, optional
             Size of a frame, in bytes.  Required if `template_header` is
-            ``None``.
+            `None`.
```
not given

mhvk commented on baseband/mark5b/base.py:
```diff
@@ -172,40 +172,40 @@ def write_frame(self, data, header=None, bps=2, valid=True, **kwargs):
 
-    This wrapper is allows one to access a Mark 5B file as a continues series
-    of samples.  Note that possible gaps in the data stream are not filled in.
+    Allows access a Mark 5B file as a continues series of samples.  Note that
```
I'm actually not sure why we mention the gaps here and nowhere else. Would suggest to remove it.

mhvk commented on baseband/mark5b/base.py:
```diff
@@ -172,40 +172,40 @@ def write_frame(self, data, header=None, bps=2, valid=True, **kwargs):
     ref_time : `~astropy.time.Time`, or None, optional
         Reference time within 500 days of the observation start time, used
-        to infer the full MJD.  Only used if `kday` is ``None``.
+        to infer the full MJD.  Only used if `kday` is `None`.
```
not given

mhvk commented on baseband/mark5b/base.py:
```diff
@@ -308,45 +308,47 @@ def _read_frame(self, fill_value=0.):
+        If no header is give, an attempt is made to construct one from these.
+        For a standard header, the following suffices.
 
     --- Header kwargs : (see :meth:`~baseband.mark5b.Mark5BHeader.fromvalues`)
```
I like your phrasing elsewhere as `keywords`

mhvk commented on baseband/mark5b/base.py:
```diff
@@ -308,45 +308,47 @@ def _read_frame(self, fill_value=0.):
-        Sets bcd-encoded unit day, hour, minute, second, and fraction, as
-        well as the frame number.
+        Start time of the file.  Sets bcd-encoded unit day, hour, minute,
+        second, and fraction, as well as the frame number, in the header.
```
Better make sure this is true...

mhvk commented on baseband/mark5b/base.py:
```diff
@@ -402,40 +404,40 @@ def write(self, data, invalid_data=False):
-    the full MJD.  Only used if `kday` is ``None``.
-subset : slice or other indexing object, optional
-    Specific components (ie. channels) of the complete sample to decode.  By
+    the full MJD.  Only used if `kday` is `None`.
```
not given

mhvk commented on baseband/mark5b/frame.py:
```diff
@@ -69,24 +69,24 @@ def __init__(self, header, payload, valid=None, verify=True):
         ref_time : `~astropy.time.Time`, or None, optional
             Reference time within 500 days of the observation time, used to
-            infer the full MJD.  Used only if `kday` is ``None``.
+            infer the full MJD.  Used only if ``kday`` is `None`.
```
not given

mhvk commented on baseband/mark5b/header.py:
```diff
@@ -43,14 +43,14 @@ class Mark5BHeader(VLBIHeaderBase):
     ref_time : `~astropy.time.Time`, or None, optional
         Reference time within 500 days of the observation time, used to infer
-        the full MJD.  Used only if `kday` is ``None``.
+        the full MJD.  Used only if ``kday`` is `None`.
```
not given

mhvk commented on baseband/vdif/base.py:
```diff
@@ -277,29 +277,29 @@ def __repr__(self):
-    samples.  Invalid data are marked, but possible gaps in the data stream
-    are not yet filled in.
+    Allows access to a VDIF file as a continuous series of samples.  Invalid
+    data are marked, but possible gaps in the data stream are not yet filled
```
As for Mark 5B, maybe just remove what we do not yet do. In fact, even remove the invalid part? That will soon be covered by `fill_value` anyway.

mhvk commented on baseband/vdif/base.py:
```diff
@@ -430,57 +431,62 @@ def _read_frame_set(self):
         Whether data is complex.
     bps : int
-        Bits per sample (or real, imaginary component).
+        Bits per elementary sample (i.e. real or imaginary part of a complex
```
as above.

mhvk commented on baseband/vdif/base.py:
```diff
@@ -430,57 +431,62 @@ def _read_frame_set(self):
         ``frame_length``, the number of long words for header plus payload.
         For some EDV, this number is fixed (e.g., ``frame_length=629`` for
         ``edv=3``, which corresponds to 20000 real 2-bit samples per frame).
     station : 2 characters
```
Should this be `str or int`?

mhvk commented on baseband/vdif/base.py:
```diff
@@ -543,36 +549,39 @@ def write(self, data, invalid_data=False):
 
 open = make_opener('VDIF', globals(), doc="""
---- For reading : (see :class:`VDIFStreamReader`)
+--- For reading a stream : (see :class:`VDIFStreamReader`)
```
Does this link work? Elsewhere, we spell out `~baseband.vdif.base.VDIFStreamReader`

mhvk commented on baseband/vdif/base.py:
```diff
@@ -543,36 +549,39 @@ def write(self, data, invalid_data=False):
     decoded data.
 
---- For writing : (see :class:`VDIFStreamWriter`)
+--- For writing a stream : (see :class:`VDIFStreamWriter`)
```
as above

mhvk commented on baseband/vlbi_base/payload.py:
```diff
@@ -27,11 +27,11 @@ class VLBIPayloadBase(object):
+    sample_shape : tuple
+        Shape of the samples; e.g., (nchan,).  Default: ().
     bps : int
         Number of bits per sample part (i.e., per channel and per real or
```
Use the standard "elementary sample" explanation here.

cczhu commented on baseband/dada/base.py:
```diff
@@ -389,35 +372,35 @@ def _get_frame(self, frame_nr):
 
-header : `~baseband.dada.DADAHeader`, optional
-    Header for the first frame, holding time information, etc.
+header : `~baseband.dada.DADAHeader`
```
I don't want to write that the sample rate is optional, despite having a default of `None`, and having an optional entry followed by a non-optional one seemed strange.

cczhu commented on baseband/gsb/base.py:
```diff
@@ -176,40 +176,42 @@ def __repr__(self):
         ``payloadsize`` is set to 2**22 (4 MB) for rawdump, and 2**23 (8 MB)
         divided by the number of streams per polarization for phased.
+    nchan : int, optional
+        Number of channels. Default: `None`, which sets it to 1 for rawdump,
```
In general, I've written descriptions for arguments even if they seem self-explanatory (better to seem pedantic than confusing).  It also sort of ties in to the glossary, though I don't think explicitly linking the docstrings to the blbliography (if even possible) is a good idea.

mhvk replied:
I think I was not clear here: I meant to remove "`None`, which sets it to`, so that one ends up with `Number of channels. Default: 1 for ...` - I feel this is closer to the style of the other descriptions, and I don't feel there is a sense of noting that the default is indicated by `None` (i.e., I wouldn't want people to pass that in).

cczhu commented on baseband/dada/base.py:
```diff
@@ -389,35 +372,35 @@ def _get_frame(self, frame_nr):
+    Start time of the file.
 offset : `~astropy.units.Quantity`
     A possible time offset from the start of the whole observation.
 npol : int, optional
```
Better to remove `optional`, since they're all individually optional, but some combination of them is necessary...

cczhu replied:
On further thought, there are necessary keywords with no defaults, so I'm keeping "optional", and reordering the keywords so the essential ones go first.

cczhu commented on baseband/gsb/base.py:
```diff
@@ -66,12 +66,12 @@ def read_payload(self, payloadsize, nchan=1, bps=4, complex_data=False):
         bps : int
-            Number of bits per sample (or real/imaginary component).
-            Default: 4.
+            Bits per elementary sample (i.e. real or imaginary component of a
```
Brought it out of the parenthesis, and reworded slightly to reduce the number of commas.

cczhu commented on baseband/gsb/base.py:
```diff
@@ -355,80 +357,82 @@ def _read_frame(self, fill_value=0., out=None):
-        `gps`.)
+        Start time of the file.  Can instead pass a timestamp as a string of
+        format ``'YYYY MM DD HH MM SS 0.SSSSSSSSS'`` to the key `gps`.
     pc_time : `~astropy.time.Time`
```
Weirdly the defaults for frame number and `mem_block` are both 1.  I've changed the latter to 0, bu all the timestamp files I have start at something like frame 20000.

cczhu commented:
This is possibly pedantic, but I would argue that "is not given" and "is `None`" describe two separate behaviours.  For the former, the code would search, say, kwargs for a key, and if it doesn't exist, would default to some behaviour, and for the latter, the code would respond to `value=None`, which happens to be the default.  For the latter the user may, for example, neglect to pass `bps` or explicitly pass `bps=None` to `gsb.open`, and in both cases it will set it to 4 or 8 depending on the circumstances.

cczhu commented:
Just recording down our discussion:

- "If `None`" is used in cases where `None` actually does something.  "If not given" is used in cases where the user must pass one term or the other, such as `ref_time` and `decade`.
- "optional" is not used for terms where the user must pass one term or another.
- Header keywords will use "optional" for any key that has a meaningful default (eg. `nchan` for DADA).

cczhu commented:
Pushing some incomplete changes...

cczhu commented:
Rebased the PR on all the changes we've made recently, and also addressed the PR comments.  Made some additional name and docstring changes so Baseband's docs are consistent.

Rules followed for naming and documentation:

* Whether an argument is classified as "optional" depends on whether the user must provide alternative arguments or keywords if they don't provide it.  In those cases, "is not given" is used.  If the code has a way of setting the value (or ignoring it altogether) if the user doesn't pass it, the argument is optional, a default is mentioned, and "is None" is used when appropriate.
* `header` in the stream writer classes is now `header0`.  This keeps the classes consistent with those in `vlbi_base`, and is more descriptive of what the user is actually passing.
* If an argument type is a Baseband, Numpy or Astropy class, it is linked (eg. `~baseband.vdif.VDIFFrameset`).  If it is a Python type or built-in (eg. `int`, `None`), it is not.  If an argument or output is a class instance, the word "instance" is not used unless not doing so leads to confusion.
* There are three ways to indicate a default: "Default: X.", "...(default: X)...", and "...X (default)...".
* Header keywords will use "optional" for any key that has a meaningful default (eg. nchan for DADA).
* Added the full path for all Baseband classes.  I've noticed linking is inconsistent if the full path is not given.
* In some cases, separated a docstring into `Parameters` and `Notes` so that Sphinx properly handles the additional text.
* Replaced LICENSE.rst with LICENSE.

Notes/Known Issues:
* Consistency between class and property docstrings hasn't yet been checked.
* For some of the GSB arguments like nchan, should we do something like "If `None` (default), 1 for rawdump..."?  It's more consistent with the standard above, but the user really shouldn't be touching nchan, for example, to deliberately set them to 1 if rawdump, 512 if phased.
* Not sure why in some cases not including the full path to a class will still properly link it in the docs.  Right now I just use what works, but am happy with always specifying the full path if you think it would be helpful to non-readthedocs readers.
* We don't use single or double backquotes consistently when referencing other arguments in docstrings.

mhvk commented on baseband/dada/base.py:
```diff
@@ -141,14 +141,15 @@ def write_frame(self, data, header=None, **kwargs):
+            if ``data`` is a `~baseband.dada.DADAFrame` instance.
         **kwargs
-            Used to initialize a header if none was passed in explicitly.
+            If `header` is not given, these are used to initialize one.
```
Answering myself: this does not produce a link, but rather typesets it "just right". Better than double quotes.

cczhu replied:
I don't think it's possible for Sphinx to link to other arguments unless they're properties with their own docstrings.  For arguments, double and single backquotes result in identical typesets.

I think in general we should reserve double backquotes for literal blocks, and single quotes for variables.

mhvk commented:
@cczhu - I'm  going ahead and just make my requested changes.

mhvk commented on baseband/dada/base.py:
```diff
@@ -251,20 +238,22 @@ def _read_frame(self, index):
+
     Parameters
     ----------
     raw : filehandle
```
Leaving this for my other PR.

mhvk commented on baseband/gsb/base.py:
```diff
@@ -176,45 +178,46 @@ def __repr__(self):
         Number of bytes per payload, divided by the number of raw files.
-        If both ``samples_per_frame`` and ``payloadsize`` are ``None``,
+        If both ``samples_per_frame`` and ``payloadsize`` are `None`,
         ``payloadsize`` is set to 2**22 (4 MB) for rawdump, and 2**23 (8 MB)
```
leaving this for another PR - not quite sure anymore what is appropriate...

mhvk commented on baseband/gsb/payload.py:
```diff
@@ -59,16 +60,15 @@ class GSBPayload(VLBIPayloadBase):
-    sample_shape : tuple
+    bps : int, optional
+        Bits per elementary sample.  Default: 2.
+    sample_shape : tuple, optional
```
And in fact with `vlbi_base`, so changed.
